dask/array/backends.py

```diff
@@ -203,3 +219,77 @@ def _tensordot_scipy_sparse(a, b, axes):
         return a * b
     elif a_axis == 1 and b_axis == 1:
         return a * b.T
+
+
+@numel_lookup.register(np.ma.masked_array)
+def _numel_masked(x, **kwargs):
+    """Numel implementation for masked arrays."""
+    return chunk.sum(np.ones_like(x), **kwargs)
+
+
+@numel_lookup.register((object, np.ndarray))
+def _numel_ndarray(x, **kwargs):
+    """Numel implementation for arrays that want to return numel of type ndarray."""
+    return _numel(x, coerce_ndarray=True, **kwargs)
+
+
+def _numel_arraylike(x, **kwargs):
+    """Numel implementation for arrays that want to return numel of the same type."""
+    return _numel(x, coerce_ndarray=False, **kwargs)
+
+
+def _numel(x, coerce_ndarray: bool, **kwargs):
+    """
+    A reduction to count the number of elements.
+
+    This has an additional kwarg in coerce_ndarray, which determines
+    whether to ensure that the resulting array is an ndarray, or whether
+    we allow it to be other array types via `np.full_like`.
+    """
+    shape = x.shape
+    keepdims = kwargs.get("keepdims", False)
+    axis = kwargs.get("axis", None)
+    dtype = kwargs.get("dtype", np.float64)
+
+    if axis is None:
+        prod = np.prod(shape, dtype=dtype)
+        if keepdims is False:
+            return prod
+
+        if coerce_ndarray:
+            return np.broadcast_to(np.array(prod, dtype=dtype), shape=(1,) * len(shape))
```

Comment:
I'm curious about the use of `np.broadcast_to` here. Naively I would have thought we could convert `prod` to an ndarray and then passed it though to the `np.full_like(...)`  call in the `else:` block below (though I could also be missing something)

Reply:
I suppose you're right, this `np.broadcast_to` isn't necessary, since we are just returning a single-value (but N-D) array. I think it's more important in the next invocation, which avoids the need to allocate a new dense array that just has a single value in it. If I understand correctly, you are suggesting something like

```python
if coerce_ndarray:
    x = np.ndarray(prod, dtype=dtype)
return np.full_like(x, prod, shape=(1,) * len(shape), dtype=dtype)
```
In that case, the new `x` won't really be used either, except to note that it's an `ndarray`.

Could also just do
```python
if coerce_ndarray:
    return np.full(prod, shape=(1,)*len(shape), dtype=dtype)
```

Reply:
What is the _real_ intent of `coerce_ndarray` flag? The docstring reads:

>    This has an additional kwarg in coerce_ndarray, which determines
>    whether to ensure that the resulting array is an ndarray, or whether
>    we allow it to be other array types via `np.full_like`.

The docstring seems incorrect to me. `np.full_like` will result in an ndarray too, but it _may not be a NumPy ndarray_, it could be, for example, a Dask array or CuPy array, depending on the type of `prod`. Is the docstring and actual intent of this flag to coerce necessarily to a NumPy ndarray? If so, then I think @ian-r-rose comment above is correct, but the docstrings should be rectified to read: "... ensure that the resulting array is a **NumPy** ndarray ...`.

Reply:
Correct, I meant a `numpy.ndarray`, thanks for the prod. I'll update the docstring.